### PR TITLE
Avoid busy spin in skip-merge FileRegion transfer

### PR DESCRIPTION
**Pull request by binmahone**
## What

This changes `MultiSegmentFileRegion.transferTo()` to issue at most one `WritableByteChannel.write()` per call and return that byte count to Netty.

If the target channel is temporarily not writable and returns `0`, the method returns `0` without advancing the segment offset. Netty can then retry the same `FileRegion` later. The handle lease added by #15064 keeps the underlying shuffle handles alive across that retry interval.

## Why

The previous implementation looped while the buffer still had remaining bytes. That assumes every non-negative write makes progress. A non-blocking channel can legally return `0`; in that case the byte buffer does not advance and the executor server thread can busy-spin inside `transferTo()`.

The fix makes the `FileRegion` honor network backpressure by yielding to Netty instead of monopolizing the event-loop thread.

## Validation scope

I do not currently have a reproducible end-to-end setup that can force this socket-backpressure condition in a controlled A/B comparison. The stall was observed in a real environment, but it cannot currently be reproduced on demand. This PR therefore does not claim an end-to-end query or shuffle performance improvement. It validates the required no-progress behavior deterministically in a focused unit test.

## Tests

- Added a deterministic `MultiSegmentFileRegionSuite` regression test covering a channel that returns `0` on the first write and accepts bytes on the second write. It verifies that the first call returns `0` after one write attempt without advancing transfer state, and that a later retry transfers the complete payload.
- Updated the test for the handle-lease constructor and verified that the lease is acquired and released.
- Merged current `main` into the branch.
- `mvn -pl sql-plugin -am -Dbuildver=351 -DskipTests install`
- `mvn -pl sql-plugin -Dbuildver=351 -DwildcardSuites=com.nvidia.spark.rapids.MultiSegmentFileRegionSuite test` — 1 test passed.

### Checklists

Documentation
- [ ] Updated for new or modified user-facing features or behaviors
- [x] No user-facing change

Testing
- [x] Added or modified tests to cover new code paths
- [ ] Covered by existing tests
      (Please provide the names of the existing tests in the PR description.)
- [ ] Not required

Performance
- [ ] Tests ran and results are added in the PR description
- [ ] Issue filed with a link in the PR description
- [x] Not required

### CODE REVIEW

**Comment by binmahone** [comment]
build

**Review by amahussein — COMMENTED** [review]
This change should be merged after #15064 

This PR returns 0 to Netty and retries the same MultiSegmentFileRegion later.
During that retry gap, current main allows unregisterShuffle to close/delete the underlying SpillablePartialFileHandle.
The next retry calls readAt on that closed handle, producing the ClosedChannelException/failed-fetch behavior tracked by #15018

**Comment by nvauto** [comment]
**NOTE: release/26.08 has been created from main. Please retarget your PR to release/26.08 if it should be included in the release.**

**Comment by binmahone** [comment]
> This change should be merged after #15064
> 
> This PR returns 0 to Netty and retries the same MultiSegmentFileRegion later. During that retry gap, current main allows unregisterShuffle to close/delete the underlying SpillablePartialFileHandle. The next retry calls readAt on that closed handle, producing the ClosedChannelException/failed-fetch behavior tracked by #15018

@amahussein any progress on #15064 ?

**Comment by amahussein** [comment]
> @amahussein any progress on #15064 ?

It was blocked on reviews. I updated the PR resolving the conflicts and pinged @abellina to TAL.

**Review by abellina — DISMISSED** [review]
LGTM, just needs updating.

**Review by amahussein — CHANGES_REQUESTED** [review]
@binmahone Can you please update the branch? Is there any performance number that can be added to the PR description to see how this change improves the corner case?

**Comment by binmahone** [comment]
Updated the branch by merging current `main`, adapted the regression test to the handle-lease changes from #15064, and reran the focused test successfully.

To be transparent, I do not currently have a reproducible end-to-end setup that can force this socket-backpressure condition in a controlled A/B comparison. We observed the stall in a real environment, but cannot reliably reproduce it on demand.

The committed unit test is a deterministic regression test for the required `FileRegion` behavior: return `0` after one write attempt without advancing state, then complete the transfer correctly on a later retry.

**Comment by binmahone** [comment]
build

**Comment by binmahone** [comment]
build

**Comment by binmahone** [comment]
@res-life, could this Blossom failure be related to the suite-scheduling change in #15854?

Before updating this branch, Blossom passed on `7817f34f2`. After merging current `main`, including `576a593b6`, `mvn verify` failed only in these two `CachedBatchWriterSuite` tests:

- `convert large columnar batch to cached batch on single col table`
- `convert large columnar batch to cached batch on multi-col table`

Both fail with Mockito reporting that a `TableWriter` mock cannot be returned by `Table.writeParquetChunked()`, which expects `ParquetTableWriter`.

This PR changes only `MultiSegmentFileRegion` and its focused test, so there is no direct code overlap. I realize #15854 changes the scheduling of `ParquetWriterSuite`, not `CachedBatchWriterSuite`, but could the changed batching/order or persistent-fork reuse expose an interaction with the static `Table` mock in `CachedBatchWriterSuite`?

Blossom log: https://github.com/NVIDIA/cudf-spark/actions/runs/33761939016

**Comment by binmahone** [comment]
build

**Comment by binmahone** [comment]
build

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/MultithreadedShuffleBufferCatalog.scala [hunk 1]
```diff
@@ -566,15 +566,11 @@ class MultiSegmentFileRegion(
...
+          // Write once and let Netty retry this FileRegion when the channel is not writable.
           val writeBuffer = ByteBuffer.wrap(readBuffer, 0, bytesRead)
-          var written = 0
-          while (writeBuffer.hasRemaining) {
-            val w = target.write(writeBuffer)
-            if (w < 0) {
-              throw new IOException("Failed to write to target channel")
-            }
-            written += w
+          val written = target.write(writeBuffer)
```
Comment (amahussein) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/MultithreadedShuffleBufferCatalog.scala:571]: Non-blocking, and the material for the description's performance line. When the channel takes only part of the buffer this discards the tail, and the next call re-reads it. Measured on a 160KB payload: 1.00x reads at a full 64KB accept, identical to the old loop, and 13x at a 4KB accept. Still far cheaper than the spin it replaces.

File: sql-plugin/src/test/scala/com/nvidia/spark/rapids/MultiSegmentFileRegionSuite.scala [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+    verify(handle).acquireRead()
+    verify(handle).releaseRead()
+  }
+
+  private final class ZeroThenWriteChannel extends WritableByteChannel {
+    private val output = new ByteArrayOutputStream()
+    private var open = true
+    var writeCalls: Int = 0
+
+    override def write(src: ByteBuffer): Int = {
```
Comment (amahussein) [inline sql-plugin/src/test/scala/com/nvidia/spark/rapids/MultiSegmentFileRegionSuite.scala:74]: Non-blocking: this fake is all or nothing, so the suite never sees a short write. On a non-blocking socket that is the ordinary case and a zero write is the corner case, so the common path is the untested one. Accepting a few bytes per call and asserting the payload and `transferred()` across calls would pin the re-read arithmetic.

**Comment by amahussein** [comment]
> @res-life, could this Blossom failure be related to the suite-scheduling change in #15854?
> 
> Before updating this branch, Blossom passed on `7817f34f2`. After merging current `main`, including `576a593b6`, `mvn verify` failed only in these two `CachedBatchWriterSuite` tests:
> 
> * `convert large columnar batch to cached batch on single col table`
> * `convert large columnar batch to cached batch on multi-col table`
> 
> Both fail with Mockito reporting that a `TableWriter` mock cannot be returned by `Table.writeParquetChunked()`, which expects `ParquetTableWriter`.
> 
> This PR changes only `MultiSegmentFileRegion` and its focused test, so there is no direct code overlap. I realize #15854 changes the scheduling of `ParquetWriterSuite`, not `CachedBatchWriterSuite`, but could the changed batching/order or persistent-fork reuse expose an interaction with the static `Table` mock in `CachedBatchWriterSuite`?
> 
> Blossom log: https://github.com/NVIDIA/cudf-spark/actions/runs/33761939016

That one was #15888, not #15854. cuDF narrowed the static `Table.writeParquetChunked` overloads to return `ParquetTableWriter`, which broke the static mock in that suite. It is closed, and premerge 14392 is green on your current head, so this is moot now.

**Review by amahussein — APPROVED** [review]
LGTM.

The Jul 7 condition is discharged: #15064 is merged, this branch has it, and `MultiSegmentFileRegion.deallocate` closes the lease on every path including a thrown write, so returning zero across a retry gap is safe now.

The single write is the right shape. Netty treats a zero return as a full send buffer, registers `OP_WRITE` and stops, and Spark's own `MessageWithHeader` documents in its class comment that a `FileRegion` may need several `transferTo` calls "to avoid busy waiting". The old inner loop broke that contract. `MultiSegmentFileRegionSuite` passes here on buildver 353, and premerge 14392 is green on this head.

On the Performance checklist I am not blocking, but `Not required` is not right either: the change does alter runtime work under backpressure. Two sentences in the description saying what it costs would settle it, and I left the measured numbers inline. Filing the field-stall bug and linking it would work too, and would fix the missing `Fixes` line, which the template asks for and this description dropped.

Two non-blocking comments inline. Neither needs another round from me.

**Comment by binmahone** [comment]
build

**Comment by amahussein** [comment]
build

**Comment by binmahone** [comment]
build

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/MultithreadedShuffleBufferCatalog.scala (hunk 1 above)
Comment (binmahone) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/MultithreadedShuffleBufferCatalog.scala:592]: ok, I'll resolve it first to get PR in

File: sql-plugin/src/test/scala/com/nvidia/spark/rapids/MultiSegmentFileRegionSuite.scala (hunk 2 above)
Comment (binmahone) [reply sql-plugin/src/test/scala/com/nvidia/spark/rapids/MultiSegmentFileRegionSuite.scala:74]: ok, I'll resolve it first to get PR in
